snow/engine/snowman/bootstrap/bootstrapper.go

```diff
@@ -566,168 +555,128 @@ func (b *Bootstrapper) markUnavailable(nodeID ids.NodeID) {
+				numFetched-b.initiallyFetched,         // Number of blocks we have fetched during this run
 				totalBlocksToFetch-b.initiallyFetched, // Number of blocks we expect to fetch during this run
 			)
 			b.fetchETA.Set(float64(eta))
 
 			if !b.restarted {
 				b.Ctx.Log.Info("fetching blocks",
-					zap.Uint64("numFetchedBlocks", blocksFetchedSoFar),
+					zap.Uint64("numFetchedBlocks", numFetched),
 					zap.Uint64("numTotalBlocks", totalBlocksToFetch),
 					zap.Duration("eta", eta),
 				)
 			} else {
 				b.Ctx.Log.Debug("fetching blocks",
-					zap.Uint64("numFetchedBlocks", blocksFetchedSoFar),
+					zap.Uint64("numFetchedBlocks", numFetched),
 					zap.Uint64("numTotalBlocks", totalBlocksToFetch),
 					zap.Duration("eta", eta),
 				)
 			}
 		}
+	}
 
-		// Attempt to traverse to the next block
-		parentID := blk.Parent()
-
-		// First check if the parent is in the processing blocks set
-		parent, ok := processingBlocks[parentID]
-		if ok {
-			blk = parent
-			continue
-		}
-
-		// If the parent is not available in processing blocks, attempt to get
-		// the block from the vm
-		parent, err = b.VM.GetBlock(ctx, parentID)
-		if err == nil {
-			blk = parent
-			continue
-		}
-		// TODO: report errors that aren't `database.ErrNotFound`
-
-		// If the block wasn't able to be acquired immediately, attempt to fetch
-		// it
-		b.Blocked.AddMissingID(parentID)
-		if err := b.fetch(ctx, parentID); err != nil {
-			return err
-		}
-
-		if err := b.Blocked.Commit(); err != nil {
-			return err
-		}
-		return b.tryStartExecuting(ctx)
+	if err := batch.Write(); err != nil || !foundNewMissingID {
+		return err
 	}
+
+	b.missingBlockIDs.Add(missingBlockID)
+	// Attempt to fetch the newly discovered block
+	return b.fetch(ctx, missingBlockID)
 }
 
 // tryStartExecuting executes all pending blocks if there are no more blocks
 // being fetched. After executing all pending blocks it will either restart
 // bootstrapping, or transition into normal operations.
 func (b *Bootstrapper) tryStartExecuting(ctx context.Context) error {
-	if numPending := b.Blocked.NumMissingIDs(); numPending != 0 {
+	if numMissingBlockIDs := b.missingBlockIDs.Len(); numMissingBlockIDs != 0 {
 		return nil
 	}
 
 	if b.Ctx.State.Get().State == snow.NormalOp || b.awaitingTimeout {
 		return nil
 	}
 
-	if !b.restarted {
-		b.Ctx.Log.Info("executing blocks",
-			zap.Uint64("numPendingJobs", b.Blocked.PendingJobs()),
-		)
-	} else {
-		b.Ctx.Log.Debug("executing blocks",
-			zap.Uint64("numPendingJobs", b.Blocked.PendingJobs()),
-		)
+	lastAcceptedHeight, err := b.getLastAcceptedHeight(ctx)
+	if err != nil {
+		return err
 	}
 
-	executedBlocks, err := b.Blocked.ExecuteAll(
-		ctx,
-		b.Config.Ctx,
-		b,
-		b.restarted,
-		b.Ctx.BlockAcceptor,
+	log := b.Ctx.Log.Info
+	if b.restarted {
+		log = b.Ctx.Log.Debug
+	}
+
+	numToExecute := b.tree.Len()
+	err = execute(
+		&haltableContext{
+			Context:  ctx,
+			Haltable: b,
+		},
```

Comment:
I'm not sure that this is safe. This context is passed into `Verify` and `Accept` - which could cause errors that were thought to be FATAL by the VMs to be treated non-fatally by the engine. It would probably be better to just pass in the haltable reference into `execute` for now.